net_state: Remove `prop_list`

cathay4t opened the pull request:
The original reason of introducing `prop_list` is to indicate which
value is defined by user instead of default value or merged value.
This is hack which require a lot of documentation and human effort on
code review.

We do not need to do this hack any more as we have MergedNetworkState
which contains 3 states now:
 * Desired state
 * Current state
 * Merged  state

For `NetworkState`, only `ovsdb`, `dns`, `hostname` do not support
partial editing, hence all of them are protected by `Option<T>` to
indicate whether user desired or not.

cathay4t commented:
Since this is big impact changes, please only merge it after QE feedback.

cathay4t commented:
> Since this is big impact changes, please only merge it after QE feedback.

I got positive feedback from our QE, let's postpone this patch to next-next release. The coming release is supposed to be a quick fix for ipsec.

qinqon commented on rust/src/lib/ovs.rs:
```diff
@@ -33,7 +32,7 @@ impl OvsDbGlobalConfig {
     // User want to remove all settings except OVN.
     pub(crate) fn is_purge(&self) -> bool {
-        self.prop_list.is_empty()
+        self.external_ids.is_none() && self.other_config.is_none()
```
How do we express this at yaml ? I think previously it was something like `external-ids: {}` so this breaks compatiblity.

cathay4t replied:
Fixed and also included unit test cases for this.

To purge ovs-db config, you may:

```yaml
---
ovs-db: {}
```

or

```yaml
---
ovs-db:
  external_ids: {}
  other_config: {}
```

qinqon reviewed (changes requested):
One question related to explicitly removed ovs stuff.

cathay4t committed:
net_state: Remove `prop_list`

The original reason of introducing `prop_list` is to indicate which
value is defined by user instead of default value or merged value.
This is hack which require a lot of documentation and human effort on
code review.

We do not need to do this hack any more as we have MergedNetworkState
which contains 3 states now:
 * Desired state
 * Current state
 * Merged  state

For `NetworkState`, only `ovsdb`, `dns`, `hostname` do not support
partial editing, hence all of them are protected by `Option<T>` to
indicate whether user desired or not.

Signed-off-by: Gris Ge <fge@redhat.com>

liangwen12year commented:
Sorry for commenting late, I reviewed it yesterday, and tested it this morning, I did not find any problem. 

Compared to before,  I noticed that some fields are skipping serializing after applying if not specified by user. such as `dns-resolver: {}`, `ovs-db: {}` (I think that is expected as we do not allow partial editing on those fields), but when we purge those settings (via `ovs-db: {}`),  I can still see it:
```
route-rules: {}
routes: {}
interfaces: []
ovs-db: {}
ovn: {}
```

 LGTM.

cathay4t commented:
I cannot understand `some fields are skipping serializing after applying if not specified by user`.

Can you provide example or be more specific?

liangwen12year commented:
> I cannot understand `some fields are skipping serializing after applying if not specified by user`.
> 
> Can you provide example or be more specific?

It is more like a comment than a concern.

After applying, `dns-resolver: {}`, `ovs-db: {}` etc are omitted if not specified by user:

```
route-rules: {}
routes: {}
interfaces:
- name: br0
  type: ovs-bridge
  state: up
  bridge:
    port:
    - name: ovs0
    - name: bond1
      link-aggregation:
        mode: balance-slb
        port:
        - name: eth2
        - name: eth1
ovn: {}
```

but when we purge the setting (dns for example), we can see that from the result below that the dns-resolver config was purged,  

```
[root@fcvm rust]# ./target/debug/nmstatectl  apply test3.yml  
[2024-01-25T09:57:12Z INFO  nmstate::query_apply::net_state] Created checkpoint /org/freedesktop/NetworkManager/Checkpoint/11
[2024-01-25T09:57:12Z INFO  nmstate::query_apply::net_state] Destroyed checkpoint /org/freedesktop/NetworkManager/Checkpoint/11
dns-resolver:
  config: {}
route-rules: {}
routes: {}
interfaces: []
ovn: {}
```
